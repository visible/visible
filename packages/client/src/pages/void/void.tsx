import * as UI from '@visi/ui';
import React from 'react';
import Helmet from 'react-helmet';
import { useTranslation } from 'react-i18next';

export const Void = () => {
  const { t } = useTranslation();

  const title = t('void.title', 'You hit the void!');
  const description = t(
    'void.description',
    'The page you were looking for was not found',
  );

  return (
    <UI.Content>
      <Helmet>
        <title>{title}</title>
        <meta name="description" content={description} />
        <meta property="og:title" content={title} />
        <meta property="og:description" content={description} />
      </Helmet>

      <h1>{t('void.title', 'You hit the void!')}</h1>
      <p>{description}</p>
    </UI.Content>
  );
};
