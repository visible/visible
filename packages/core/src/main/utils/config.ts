import { Config } from '../domain/config';

/**
 * Takes config object and reoslve extends
 * @param baseConfig base configuration object
 */
export function mergeExtends(baseConfig: Config) {
  const extendables: Config[] = [];

  for (const extendable of baseConfig.extends) {
    const config = require(extendable);
    extendables.push(config);
  }

  return extendables.reduce((result, config) => {
    return { ...result, ...config };
  }, baseConfig);
}
