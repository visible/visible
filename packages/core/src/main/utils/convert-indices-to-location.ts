import { Location } from '../../shared';

const convertIndexToLocation = (
  source: string,
  position: number,
): [number, number] => {
  const lines = source.substr(0, position).split('\n');
  const ancestorLines = lines.slice(0, lines.length - 1);

  const line = lines.length;
  const column = position - ancestorLines.join('\n').length;

  return [line, column];
};

export const convertIndicesToLocation = (
  source: string,
  startIndex: number,
  endIndex: number,
): Location => {
  const startsAt = convertIndexToLocation(source, startIndex);
  const endsAt = convertIndexToLocation(source, endIndex);

  return {
    startLine: startsAt[0],
    startColumn: startsAt[1],
    endLine: endsAt[0],
    endColumn: endsAt[1],
  };
};
