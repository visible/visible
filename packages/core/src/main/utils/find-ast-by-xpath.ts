import { Element, Node } from 'domhandler';
import { parseDOM } from 'htmlparser2';
import XPathAnalyzer, {
  ABSOLUTE_LOCATION_PATH,
  EQUALITY,
  FUNCTION_CALL,
  NODE_NAME_TEST,
  NODE_TYPE_TEST,
  NUMBER,
  POSITION,
  PredicateNode,
  StepNode,
} from 'xpath-analyzer';

const isSameKind = (a: Node, b: Node) => {
  if (a instanceof Element && b instanceof Element) {
    return a.tagName === b.tagName;
  }

  if (!(a instanceof Element) && !(b instanceof Element)) {
    return true;
  }

  return false;
};

const filterSameKind = (topic: Node, nodes: Node[]) =>
  nodes.filter(node => isSameKind(topic, node));

const testPredicates = (predicates: PredicateNode[], node: Node) =>
  predicates.every(predicate => {
    if (
      predicate.type === EQUALITY &&
      predicate.lhs.type === FUNCTION_CALL &&
      predicate.lhs.name === POSITION &&
      predicate.rhs.type === NUMBER &&
      node.parent
    ) {
      const selfIndex = filterSameKind(node, node.parent.childNodes).findIndex(
        child => child === node,
      );

      return selfIndex === predicate.rhs.number - 1;
    }

    // eslint-disable-next-line
    console.warn(`Predicate ${predicate.type} is not implemented yet`);
    return false;
  });

const testNodeType = (
  node: Node,
  name: string,
  predicates: PredicateNode[] = [],
) => {
  // node.nodeType is number for some reason...
  if (
    name === 'text' &&
    !(node instanceof Element) &&
    testPredicates(predicates, node)
  ) {
    return true;
  }

  return false;
};

const testNodeName = (
  node: Node,
  name: string,
  predicates: PredicateNode[] = [],
) => {
  if (
    node instanceof Element &&
    node.tagName.toLowerCase() === name &&
    testPredicates(predicates, node)
  ) {
    return true;
  }

  return false;
};

const findNode = (
  node: Node,
  [step, ...restSteps]: StepNode[],
): Node | undefined => {
  if (step == null) return node;

  if (!(node instanceof Element)) {
    throw new Error('node must be an element');
  }

  const { test } = step;

  const match = node.children.find(child => {
    switch (test.type) {
      case NODE_NAME_TEST:
        return testNodeName(child, test.name, step.predicates);
      case NODE_TYPE_TEST:
        return testNodeType(child, test.name, step.predicates);
      default:
        // eslint-disable-next-line
        console.warn(`Test type ${test.type} is not implemented yet`);
        return false;
    }
  });
  if (match == null) return;

  return findNode(match, restSteps);
};

export const findASTByXPath = (
  source: string,
  xpath: string,
): Node | undefined => {
  const htmlElement = parseDOM(source, {
    withEndIndices: true,
    withStartIndices: true,
  }).find(node => node instanceof Element && node.name === 'html');

  if (htmlElement == null) {
    throw new Error('Top-level html element were not provided');
  }

  const rootXPath = new XPathAnalyzer(xpath).parse();
  if (rootXPath.type !== ABSOLUTE_LOCATION_PATH) {
    throw new Error('Given Xpath must start with /, got' + xpath);
  }

  // Drop /html because we know we have
  const [, ...steps] = rootXPath.steps;
  const node = findNode(htmlElement, steps);

  return node;
};
