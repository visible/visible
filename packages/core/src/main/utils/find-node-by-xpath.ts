import { Protocol } from 'devtools-protocol/types/protocol';
import { CDPSession } from 'puppeteer';
import XPathAnalyzer, {
  ABSOLUTE_LOCATION_PATH,
  EQUALITY,
  FUNCTION_CALL,
  NODE_NAME_TEST,
  NODE_TYPE_TEST,
  NUMBER,
  POSITION,
  PredicateNode,
  StepNode,
} from 'xpath-analyzer';

const isSameKind = (a: Protocol.DOM.Node, b: Protocol.DOM.Node) => {
  if (a.nodeType === 1 && b.nodeType === 1) {
    return a.localName === b.localName;
  }

  if (a.nodeType === 3 && b.nodeType === 3) {
    return true;
  }

  return false;
};

const filterSameKind = (topic: Protocol.DOM.Node, nodes: Protocol.DOM.Node[]) =>
  nodes.filter(node => isSameKind(topic, node));

const testPredicates = (
  predicates: PredicateNode[],
  node: Protocol.DOM.Node,
  parent?: Protocol.DOM.Node,
) => {
  return predicates.every(predicate => {
    if (
      predicate.type === EQUALITY &&
      predicate.lhs.type === FUNCTION_CALL &&
      predicate.lhs.name === POSITION &&
      predicate.rhs.type === NUMBER &&
      parent &&
      parent.children
    ) {
      const selfIndex = filterSameKind(node, parent.children).findIndex(
        child => child.nodeId === node.nodeId,
      );
      return selfIndex === predicate.rhs.number - 1;
    }

    // eslint-disable-next-line
    console.warn(`Predicate ${predicate.type} is not implemented yet`);
    return false;
  });
};

const testNodeName = (
  node: Protocol.DOM.Node,
  name: string,
  parent?: Protocol.DOM.Node,
  predicates: PredicateNode[] = [],
) => {
  if (node.localName === name && testPredicates(predicates, node, parent)) {
    return true;
  }

  return false;
};

const testNodeType = (
  node: Protocol.DOM.Node,
  type: string,
  parent?: Protocol.DOM.Node,
  predicates: PredicateNode[] = [],
) => {
  if (
    type === 'text' &&
    node.nodeName === '#text' &&
    testPredicates(predicates, node, parent)
  ) {
    return true;
  }
};

const findNode = (
  node: Protocol.DOM.Node,
  [step, ...restSteps]: StepNode[],
): Protocol.DOM.Node | undefined => {
  if (step == null) return node;
  if (node.children == null) throw new Error('node has no children');

  const { test } = step;

  const match = node.children.find(child => {
    switch (test.type) {
      case NODE_NAME_TEST:
        return testNodeName(child, test.name, node, step.predicates);
      case NODE_TYPE_TEST:
        return testNodeType(child, test.name, node, step.predicates);
      default:
        // eslint-disable-next-line
        console.warn(`Test type ${test.type} is not implemented yet`);
        return false;
    }
  });
  if (!match) return;

  return findNode(match, restSteps);
};

export const findNodeByXPath = async (client: CDPSession, xpath: string) => {
  const { root } = await client.send('DOM.getDocument', { depth: -1 });

  const rootXPath = new XPathAnalyzer(xpath).parse();
  if (rootXPath.type !== ABSOLUTE_LOCATION_PATH) {
    throw new Error('Given Xpath must start with /, got' + xpath);
  }

  return findNode(root, rootXPath.steps);
};
