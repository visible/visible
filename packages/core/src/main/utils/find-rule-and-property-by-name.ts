import { Protocol } from 'devtools-protocol/types/protocol';
import { CDPSession } from 'puppeteer';

const findPropertyInStyle = (
  style: Protocol.CSS.CSSStyle,
  propertyName: string,
) => {
  return style.cssProperties
    .filter(property => !property.disabled && !property.implicit)
    .find(property => property.name === propertyName);
};

const flattenStylesFromMatchedStyles = (
  matched: Protocol.CSS.GetMatchedStylesForNodeResponse,
): Protocol.CSS.CSSStyle[] => {
  const {
    inlineStyle,
    attributesStyle,
    matchedCSSRules = [],
    inherited = [],
  } = matched;

  const a = matchedCSSRules
    .filter(ruleMatch => ruleMatch.rule.origin === 'regular')
    .map(ruleMatch => ruleMatch.rule.style);

  const b = inherited.flatMap(rule =>
    rule.matchedCSSRules
      .filter(ruleMatch => ruleMatch.rule.origin === 'regular')
      .map(ruleMatch => ruleMatch.rule.style),
  );

  return [inlineStyle, attributesStyle, a, b]
    .flat()
    .filter((style): style is Protocol.CSS.CSSStyle => style != null);
};

export const findRuleAndPropertyByName = async (
  client: CDPSession,
  propertyName: string,
  node: Protocol.DOM.Node,
) => {
  const matchedStyles = await client.send('CSS.getMatchedStylesForNode', {
    nodeId: node.nodeId,
  });

  const styles = flattenStylesFromMatchedStyles(matchedStyles);

  // Find style such that propertyName can be found
  const matchedStyle = styles.find(
    style => findPropertyInStyle(style, propertyName) != null,
  );

  if (matchedStyle == null) {
    throw new Error(`No matched rule has property ${propertyName}`);
  }

  // TODO: call `findPropertyInRule` just once
  const property = findPropertyInStyle(matchedStyle, propertyName);

  if (property == null) {
    throw new Error(`Rule has no property ${propertyName}`);
  }

  return { style: matchedStyle, property };
};
