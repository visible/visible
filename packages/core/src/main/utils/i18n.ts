import i18next from 'i18next';

import en from '../locale/en.json';
import ja from '../locale/ja.json';

export const createI18n = async (language?: string) => {
  const i18nextCustom = i18next.createInstance();
  const t = await i18nextCustom.init({
    lng: language,
    defaultNS: 'core',
    fallbackLng: 'en',
    resources: {
      en: { core: en },
      ja: { core: ja },
    },
  });

  return [i18nextCustom, t] as const;
};
