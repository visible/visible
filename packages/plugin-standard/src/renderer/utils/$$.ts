export const $$ = (query: string) =>
  Array.from(document.querySelectorAll(query));
