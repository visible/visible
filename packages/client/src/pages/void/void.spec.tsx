import React from 'react';

import { render } from '../../../tests/utils';
import { Void } from './void';

describe('Void', () => {
  it('matches to the snapshot', () => {
    const { container } = render(<Void />);
    expect(container).toMatchSnapshot();
  });
});
