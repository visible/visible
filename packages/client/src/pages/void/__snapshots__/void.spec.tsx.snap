// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Void matches to the snapshot 1`] = `
.c0 {
  box-sizing: border-box;
  width: 1080px;
  margin: 24px auto;
  padding: 12px 18px;
  border-radius: 12px;
  background-color: white;
  box-shadow: 0 0 24px rgba(0,0,0,0.05);
}

<div>
  <main
    class="c0"
  >
    <h1>
      You hit the void!
    </h1>
    <p>
      The page you were looking for was not found
    </p>
  </main>
</div>
`;
